lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -322,6 +322,104 @@ void SPIRVRegularizeLLVMBase::expandSYCLTypeUsing(Module *M) {
     expandVIDWithSYCLTypeByValComp(F);
 }
 
+// In this function, we handle two conversion operations
+// 1. fptoui.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// 2. fptosi.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// Such non-standard integer types cannot be handled in SPIR-V. Hence, they
+// will be promoted to
+// 1. fptoui.sat.i64.fY (Y is 32 or 64)
+// 2. fptosi.sat.i64.fY (Y is 32 or 64)
+// However, LLVM documentation requires the following rules to be obeyed.
+// Rule 1: If the argument is any NaN, zero is returned.
+// Rule 2: If the argument is smaller than the smallest representable
+// (un)signed integer of the result type, the smallest representable
+// (un)signed integer is returned.
+// Rule 3: If the argument is larger than the largest representable (un)signed
+// integer of the result type, the largest representable (un)signed integer is
+// returned.
+// Rule 4: Otherwise, the result of rounding the argument towards zero is
+// returned.
+// Rules 1 & 4 are preserved when promoting iX to i64. For preserving Rule 2
+// and Rule 3, we saturate the result of the promoted instruction based on
+// original integer type (iX)
+// Example:
+// Input:
+// %0 = call i2 @llvm.fptosi.sat.i2.f32(float %input)
+// %1 = sext i32 %0
+// Output:
+// %0 = call i32 @_Z17convert_long_satf(float %input)
+// %1 = icmp sge i32 %0, 1 <Largest 2-bit signed integer>
+// %2 = icmp sle i32 %0, -2 <Smallest 2-bit signed integer>
+// %3 = select i1 %1, i32 1, i32 %0
+// %4 = select i1 %2, i32 -2, i32 %3
+// Replace uses of %1 in Input with %4 in Output
+void SPIRVRegularizeLLVMBase::cleanupConversionToNonStdIntegers(Module *M) {
+  for (auto I = M->begin(), E = M->end(); I != E;) {
+    Function *F = &(*I++);
+    std::vector<Instruction *> ToErase;
+    for (BasicBlock &BB : *F) {
+      for (Instruction &I : BB) {
+        if (IntrinsicInst *II = dyn_cast<IntrinsicInst>(&I)) {
+          // TODO: Vector type not supported yet.
+          if (isa<VectorType>(II->getType()))
+            continue;
+          auto IID = II->getIntrinsicID();
+          auto IntBitWidth = II->getType()->getScalarSizeInBits();
+          if (IntBitWidth == 8 || IntBitWidth == 16 || IntBitWidth == 32 ||
+              IntBitWidth == 64)
+            continue;
+          if (IID == Intrinsic::fptosi_sat) {
+            // Identify sext (user of I). Make sure that's the only use of I.
+            auto *User = I.getUniqueUndroppableUser();
+            if (!User || !isa<SExtInst>(User))
+              continue;
+            auto *SExtI = dyn_cast<SExtInst>(User);
```

Comment:
`SExtI` is actually unused in the code. Can we instead check for the integer size, if it's <32 bit, then create i32 version of the emulation and if User is ext instruction - put it to remove container. Same logic for 32 < X < 64 bit range.

Reply:
> `SExtI` is actually unused in the code. Can we instead check for the integer size, if it's <32 bit, then create i32 version of the emulation and if User is ext instruction - put it to remove container. Same logic for 32 < X < 64 bit range.

Isn't this dangerous?   For example with i3, in the original code users would see values in the range [I3_MIN,I3_MAX].  With the suggested change, users will see values in the range [I32_MIN,I32_MAX].

Reply:
Why is this dangerous? [I3_MIN,I3_MAX] aren't going anywhere, it just about not checking if a 'single' user is ext instruction and doing transformation regardless.

Reply:
Thx.  I misunderstood your proposed change.

Reply:
Can you please show a IR code snippet example here? Looks like this might be a good point. But I am not fully grasping it. Thanks

Reply:
FWIW, I initially had a solution with 'not considering' if there is single user. But, in that case I end up running into issues when I try to do replaceAllUsesWith (Type of old result will be iX and new result will be i32/i64 and replaceAllUsesWith wil complain about type mismatch. There are ugly workarounds, but we can avoid by just trying to find that sext/zext user. The optimization we are trying to undo will always create this single user, so we should be ok
Tahnks

Reply:
If the sole purpose is to handle `AggressiveInstCombine` output for scalar conversions I'm OK with that patch.
If the purpose is different - please see clang's eimplementation of `ScalarExprEmitter` - it doesn't generate any integer extensions for intrinsic's result.

Reply:
Yes. handling AggressiveInstCombine is the primary purpose here. Thanks